fvtest/tril/tril/parser.cpp

```diff
@@ -0,0 +1,560 @@
+            ++tokenIt; // consume EQUALS
+            ++tokenIt; // consume to be LPAREN_SQ(the start of value list) or single node value
+            if (tokenIt->getType() == Token::LPAREN_SQ) {
+                ++tokenIt;
+                currentArg = createNodeArg(node_args_name, parseValueList(tokenIt), NULL);
+            } else {
+                try {
+                    currentArg = createNodeArg(node_args_name, buildNodeValue(*tokenIt), NULL);
+                } catch (ParserFailure &f) {
+                    std::cerr << f.what() << std::endl;
+                }
+            }
+        } else if (checkValue(tokenIt)) {
+            try {
+                currentArg = createNodeArg("", buildNodeValue(*tokenIt), NULL);
+            } catch (ParserFailure &f) {
+                std::cerr << f.what() << std::endl;
+            }
+        } else {
+            throw ParserFailure("Incorrect type for argument, current value is " + tokenIt->getValue());
+        }
+        if (argList == NULL) {
+            argList = currentArg;
+        } else {
+            appendSiblingArg(argList, currentArg);
+        }
+        ++tokenIt; // consume to be the start of the next Arg (could be single node value or ID followed by BECOMES)
+    }
+    return argList;
+}
+
+/* Parse and build the AST
+ * 
+ * @param tokenIt The tokens contains the AST info
+ * @return The root of the AST
+ */
+ASTNode* buildAST(TokenIter &tokenIt) {
+    ASTNode* siblingsList = NULL;
+    while (!tokenIt.isEnd() && tokenIt->getType() == Token::LPAREN) {
+        ASTNodeArg* argList = NULL;
+        ASTNode* currentchild = NULL;
+        ASTNode* childrenList = NULL;
+        ++tokenIt; // consume to be ID
+        if (tokenIt.isEnd()) {
+            throw ParserFailure("Token stream ended unexpectedly after LPAREN");
+        } else if (tokenIt->getType() != Token::ID) {
+            throw ParserFailure("The next token of LPAREN should be IDENTIFIER, while current next value is " +  tokenIt.peek().getValue());
+        }
+        char* node_name = new char[tokenIt->getValue().size() + 1];
+        strcpy(node_name, tokenIt->getValue().c_str());
+        ++tokenIt; // consume to be the start of argList/nodeList or RPAREN
+        if (checkValue(tokenIt)) {
+            try {
+                argList = parseArgList(tokenIt);
+            } catch (LexerFailure &f) {
+                std::cerr << f.what() << std::endl;
+            }
+        }
+        currentchild = buildAST(tokenIt);
+        if (childrenList == NULL) {
+            childrenList = currentchild;
+        } else {
+            appendSiblingNode(childrenList, currentchild);
+        }
+        if (!tokenIt.isEnd() && tokenIt->getType() != Token::RPAREN) {
+            throw ParserFailure("Missing RPAREN, while current value is " + tokenIt->getValue());
+        }
+        ++tokenIt; // consume to be the end of file or LPAREN(the sibling of the current node)
```

Comment:
If `tokenIt.isEnd()` returns true, isn't that an error as well? It means that there's a missing RPAREN, which is wrong according to the grammar.